### PR TITLE
Extend create plugin docs

### PR DESCRIPTION
**Pull request by qwerty287**
closes https://github.com/woodpecker-ci/woodpecker/issues/1389
closes https://github.com/woodpecker-ci/woodpecker/issues/1033

### CODE REVIEW

**Comment by lafriks** [comment]
It would also be nice to add a go plugin sample and it's dockerfile

**Comment by pat-s** [comment]
We should really find a central place for plugin documentation (both for dev and users). I still find it unfortunate that plugins are split between CB and GH. 

Btw there is also https://codeberg.org/woodpecker-plugins/go-plugin.

**Comment by pat-s** [comment]
> But docs are completely here, right?

I guess this would be the best.

We have a recent use case which also exactly tackles this topic: https://codeberg.org/woodpecker-plugins/node-pm/issues/21

Any effort on improving the situation is great 👍

File: docs/docs/20-usage/51-plugins/20-creating-plugins.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
+# Creating plugins
+
+Creating a new plugin is simple: Build a Docker container which uses your plugin logic as entrypoint.
```
Comment (pat-s) [inline docs/docs/20-usage/51-plugins/20-creating-plugins.md:3]:
```suggestion
Creating a new plugin is simple: Build a Docker container which uses your plugin logic as the ENTRYPOINT.
```

File: docs/docs/20-usage/51-plugins/20-creating-plugins.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+## Settings
+
+To allow users to configure the behavior of your plugin, you should use settings.
```
Comment (pat-s) [inline docs/docs/20-usage/51-plugins/20-creating-plugins.md:7]:
```suggestion
To allow users to configure the behavior of your plugin, you should use `settings:`.
```

File: docs/docs/20-usage/51-plugins/20-creating-plugins.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+These are passed to your plugin as uppercase env vars with `PLUGIN_` prefix.
```
Comment (pat-s) [inline docs/docs/20-usage/51-plugins/20-creating-plugins.md:9]:
```suggestion
These are passed to your plugin as uppercase env vars with a `PLUGIN_` prefix.
```

File: docs/docs/20-usage/51-plugins/20-creating-plugins.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+Using a setting like `url` results in `PLUGIN_URL` as env var.
```
Comment (pat-s) [inline docs/docs/20-usage/51-plugins/20-creating-plugins.md:10]:
```suggestion
Using a setting like `url` results in an env var named `PLUGIN_URL`.
```

File: docs/docs/20-usage/51-plugins/20-creating-plugins.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        list:
+          - 2
+          - 3
+```
+
+Values like this are converted to JSON and then passed to your plugin. In the example above, the env value would be `{"abc": "2", "list": [ "2", "3" ]}`.
+
+### Secrets
+
+Secrets should be passed as settings too. Therefore, users should use [`from_secret`](../40-secrets.md#use-secrets-in-settings).
```
Comment (pat-s) [inline docs/docs/20-usage/51-plugins/20-creating-plugins.md:45]:
```suggestion
Secrets should be passed as settings, too. Therefore, users should use [`from_secret`](../40-secrets.md#use-secrets-in-settings).
```

File: docs/docs/20-usage/51-plugins/20-creating-plugins.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+  -d ${PLUGIN_BODY} \
+  ${PLUGIN_URL}
+```
+
+### Package it
+
+Create a Dockerfile that adds your shell script to the image, and configures the image to execute your shell script as the main entrypoint.
+
+```dockerfile
+FROM alpine
```
Comment (pat-s) [inline docs/docs/20-usage/51-plugins/20-creating-plugins.md:88]:
```suggestion
```dockerfile
# please pin the version, e.g. alpine:3.19
FROM alpine
```

We should encourage version pinning already here

File: docs/docs/20-usage/51-plugins/20-creating-plugins.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  -e PLUGIN_BODY="hello world" \
+  foo/webhook
+```
+
+## Best practices
+
+- Build your plugin for different architectures to allow many users to use them.
+  At least, you should support `amd64` and `arm64`.
+- Provide binaries for users using the `local` backend.
+  These should also be built for different OS/architectures.
```
Comment (pat-s) [inline docs/docs/20-usage/51-plugins/20-creating-plugins.md:117]: If we list this, we should add more context how this can be done.

**Review by pat-s — COMMENTED** [review]
Two additional points worth mentioning:

- Plugins on CB are also published to the CB package registry
- `renovate` should be used to keep dependencies up-to-date
- Encourage the use of https://github.com/woodpecker-ci/plugin-ready-release-go (?)

**Comment by 6543** [comment]
i agree the docs should not be for our official plugins but rather for everybody who want to make there own.
and the programming practices might vary

**Comment by pat-s** [comment]
Programming practices is one thing but if there are no general "rules" how to format/maintain/publish/release plugins, everybody will do something else and in the end it's just a mess for users (and devs).
That's what linters are for, that's what other implicit "contributing" rules are for.

*At least* in repos which live within the GH and CB woodpecker org we should have such rules. In a central place. How devs then comply with these rules is a different thing. I am actually aiming to draft something for a long time already (content TBD) but I don't think being "too open" or "too lax" is really helping in the long-term. Yes, it's FOSS, yes individual contributors with (many) own opinions, but if this project wants to grow in a smooth way, there is no way around such rules in the mid and long-term.

> This should be added to the individual plugin docs.

I disagree. This is a perfect point showcasing what I wanted to say above: sure, it doesn't hurt anyone if a plugin is published there or if not. But if there is no rule, the ratio of plugins available there will be maybe like 70/30 (available/not available) and users have to look up every time the specific state for this plugin. This is already the case with the availability of plugins in CB or GH. There was no rule, nobody knows why the current state is like it is and users are also partly confused.
This is just one example, the same applies for 

- semver
- conventional commits
- updates
- naming
- README content
- etc.

I hope you understand what I want to say 😄 I don't want to make things too complicated and surely enjoyable for everyone. And many things are great. But there are (quite some) points where everyone is just 🤷🤷🤷 and in the end this will cause issues/lot of unnecessary typing/unprofessional look of the project.

> i agree the docs should not be for our official plugins but rather for everybody who want to make there own.

And I also disagree with this with the arguments from above.

No hard feelings, just opinions ❤️

File: docs/docs/20-usage/51-plugins/20-creating-plugins.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    image: foo/plugin
+    settings:
+      complex:
+        abc: 2
+        list:
+          - 2
+          - 3
+```
+
+Values like this are converted to JSON and then passed to your plugin. In the example above, the env value would be `{"abc": "2", "list": [ "2", "3" ]}`.
```
Comment (6543) [inline docs/docs/20-usage/51-plugins/20-creating-plugins.md:41]:
```suggestion
Values like this are converted to JSON and then passed to your plugin. In the example above, the environment variable `PLUGIN_COMPLEX` would contain `{"abc": "2", "list": [ "2", "3" ]}`.
```

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
@pat-s @qwerty287 as a owner I think this pull is ready.

And so I'm going it merge it _now_ :)

for the guidelines, I open a new discussion

**Comment by 6543** [comment]
- >  #3085

**Pull request merged**